miio/updater.py

```diff
@@ -0,0 +1,66 @@
+from http.server import HTTPServer, BaseHTTPRequestHandler
+import hashlib
+import logging
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class UpdateServer(BaseHTTPRequestHandler):
+    """A simplified handler just returning the contents of a single file."""
+    def __init__(self, request, client_address, server):
+        self.payload = server.payload
+        self.server = server
+
+        super().__init__(request, client_address, server)
+
+    def handle_one_request(self):
+        self.server.got_request = True
+        self.raw_requestline = self.rfile.readline()
+
+        if not self.parse_request():
+            print("unable to parse request")
+            return
+
+        print("got %s for %s" % (self.command, self.path))
+
+        self.send_response(200)
+        self.send_header('Content-type', 'application/octet-stream')
+        self.end_headers()
+        self.wfile.write(self.payload)
+
+
+class Updater:
+    """A simple HTTP server for serving an update file.
+
+    The server will be started in an emphemeral port, and will only accept
+    a single request to keep it simple."""
+    def __init__(self, file, interface=None):
+        addr = ('', 0)
+        self.server = HTTPServer(addr, UpdateServer)
+        self.server.got_request = False
+        _, self.port = self.server.server_address
+        _LOGGER.info("Serving on port %s" % self.port)
+        self.server.timeout = 10
+
+        with open(file, 'rb') as f:
+            self.payload = f.read()
```

Comment:
Should be memory inefficient

Reply:
I think it doesn't matter here, as the file needs to be read at some point for sending it to the client. The code is however incomplete, there is currently no way to tell interface/IP to use for the url (e.g. when one has multiple network interfaces as I do). I will have to look more into it in a week when I'll be able to test it on the device.

Reply:
The point i mean is you should read file by chunks and feed data to hash, reading hundreds megabytes to memory at once is a bad idea.

Reply:
I understand what you mean, but at some the file has to be read (again) for sending over the wire to the device. I'll look into that though, thanks :-)